File: lintcode/remove-nth-node-from-end-of-list.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/remove-nth-node-from-end-of-list/

from linked_list import *

"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The first node of linked list.
    @param n: An integer.
    @return: The head of linked list.
    """
    def removeNthFromEnd(self, head, n):
        # write your code here
        if n == 0:
            return head

        fast = head
        for i in xrange(n):
            fast = fast.next

        slow = head
        prev_slow = None
        while fast:
            prev_slow = slow
            slow = slow.next
            fast = fast.next

        if not prev_slow:
            return head.next
        else:
            prev_slow.next = slow.next
            return head

if __name__ == '__main__':
    head = create_linked_list([1,2,3,4,5,6])
    s = Solution()
    head = s.removeNthFromEnd(head, 0)
    print_linked_list(head)
    head = s.removeNthFromEnd(head, 1)
    print_linked_list(head)
    head = s.removeNthFromEnd(head, 5)
    print_linked_list(head)

File: leetcode/remove-nth-node-from-end-of-list.cc
// Question: https://leetcode.com/problems/remove-nth-node-from-end-of-list/

#include "headers.h"

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if (!head || !head->next) {
            return NULL;
        }

        ListNode* fast = head;
        int i = 0;
        while (i++ < n) {
            fast = fast->next;
        }

        ListNode dummy(0);
        dummy.next = head;
        ListNode* node = &dummy;
        while (fast) {
            node = node->next;
            fast = fast->next;
        }

        node->next = node->next->next;
        return dummy.next;
    }
};

int main(int argc, char** argv) {
    Solution s;
    ListNode* head = new ListNode(1);
    head->next = new ListNode(2);
    head->next->next = new ListNode(3);
    head->next->next->next = new ListNode(4);
    head->next->next->next->next = new ListNode(5);
    ListNode* new_head = s.removeNthFromEnd(head, 5);
    while (new_head) {
        cout << new_head->val << " ";
        new_head = new_head->next;
    }
    cout << endl;
}
